tools/src/main/java/com/nedap/archie/creation/OpenEhrRmInstanceGenerator.java

```diff
@@ -0,0 +1,375 @@
+        Object nullFlavour = result.get("null_flavour");
+        if(value == null && nullFlavour == null) {
+            Map<String, Object> dvText = new LinkedHashMap<>();
+            dvText.put(typePropertyName, "DV_TEXT");
+            dvText.put("value", "string");
+
+            result.put("value", dvText);
+        }
+        if(value != null && nullFlavour != null) {
+            result.remove("null_flavour");//can't be set at the same time
+        }
+    }
+
+    private void fixHistory(Map<String, Object> result) {
+        Object events = result.get("events");
+        if(events == null || (events instanceof List && ((List) events).isEmpty())) {
+            List<Map> newEvents = new ArrayList<>();
+            newEvents.add(this.generator.constructExampleType("EVENT"));
+            result.put("events", newEvents);
+        }
+    }
+
+    private void fixCodePhrase(Map<String, Object> result, CObject cObject) {
+        try {
+            String codeString = (String) result.get("code_string");//TODO: check terminology code to be local?
+
+            if(AOMUtils.isValueCode(codeString)) {
+                //check for OpenEHR term mapping and use that if available, so we get correct
+                //rm objects
+
+                //TODO: quite some more term ids, such as IANA characters sets, etc. HOW?
+                URI openehr = generator.getArchetype().getTerminology(cObject).getTermBinding("openehr", codeString);
+                if(openehr != null && openehr.getPath() != null) {
+                    int i = openehr.getPath().lastIndexOf('/');
+                    if (i > 0) {
+                        codeString = openehr.getPath().substring(i+1);
+                        Map<String, Object> terminologyId = (Map<String, Object>) result.get("terminology_id");
+                        if(terminologyId != null) {
+                            terminologyId.put("value", "openehr");
+                        }
+                        result.put("code_string", codeString);
+                        //TODO: add a mapping to the at code in the data.
+                    }
+                }
+            }
+        } catch (Exception e) {
+            //if statements would be cleaner, but this should not happen and is a lot less code
+            //cannot set this apparently, it will be filled by the BMM required property later
+        }
+    }
+
+    private void fixCodedText(Map<String, Object> result, CObject cObject) {
+        try {
+            Map<String, Object> definingCode = (Map<String, Object>) result.get("defining_code");
+            String codeString = (String) definingCode.get("code_string");//TODO: check terminology code to be local?
+            ArchetypeTerm term = generator.getArchetype().getTerm(cObject, codeString, generator.getLanguage());
+            result.put("value", term.getText());
+            fixCodePhrase(definingCode, cObject);
+        } catch (Exception e) {
```

Comment:
This only seems to hit NullPointerExceptions, so maybe make this more specific?
```suggestion
        } catch (NullPointerException e) {
```

Reply:
Even better, replace it with if-statements, and add a todo for some more cases that are less easy to fix, will fix.